### PR TITLE
Use atdpy

### PR DESCRIPTION
**Pull request by mjambon**
This requires the latest version of `atdpy` as specified in the opam file used to install dev dependencies. The Python code `output_from_core.py` is generated when calling `make`  (= `make build`) from the project root or when running `make build` from the Python project root (`/semgrep/`).

This branch changes a tiny bit of the Python code that interprets the output from semgrep-core. There's more to do but I want to merge this first. It will get increasingly faster as there will be less and less work to do on atdpy.

PR checklist:

- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

**Comment by mjambon** [comment]
After adding an argument to `black` in the pre-commit config so as to ignore a file, this error happens: https://github.com/returntocorp/semgrep/runs/5764698462?check_suite_focus=true

File: interfaces/Output_from_core.atd [hunk 1]
```diff
@@ -207,7 +225,8 @@ type target_time = {
   run_time: float;
 }
 
-type rule_times = {
+type rule_times
+  <python decorator="dataclass(frozen=True)"> = {
```
Comment (aryx) [inline interfaces/Output_from_core.atd:229]: maybe this dataclass(frozen=True) should be the default in atdpy? so we don't need to specify it each time?

File: semgrep/semgrep/output_from_core.py [hunk 2]
```diff
@@ -283,7 +283,7 @@ def to_json_string(self, **kw: Any) -> str:
         return json.dumps(self.to_json(), **kw)
 
 
-@dataclass
+@dataclass(frozen=True)
```
Comment (aryx) [inline semgrep/semgrep/output_from_core.py:286]: So do we want to commit the generated code in the repository? Simpler for people?

**Review by aryx — APPROVED** [review]
Awesome start!

File: interfaces/Output_from_core.atd (hunk 1 above)
Comment (mjambon) [reply interfaces/Output_from_core.atd:229]: Yes, maybe we could do that. Maybe other options would make sense by default as well.

There's the issue of overriding them vs. extending them. The default decorator is `@dataclass` but if the user specifies another `@dataclass(...)` decorator, it replaces the default. Now if the default is no longer a simple `@dataclass`, it gets tricky for the user to do the right thing.

File: semgrep/semgrep/output_from_core.py (hunk 2 above)
Comment (mjambon) [reply semgrep/semgrep/output_from_core.py:286]: I tried removing it but I got pre-commit linter errors, due to an attempt to resolve names and the absence of the generated file that provides those names. Not getting this error would mean running atdpy before running the pre-commit linters, which seemed too complicated.

File: interfaces/Output_from_core.atd [hunk 3]
```diff
@@ -223,10 +242,12 @@ type rule_times = {
 (* Synthesizing from diffs (see locate_patched_functions in Synthesizing.mli) *)
 (*****************************************************************************)
 
-type cve_result = {
+type cve_result
+  <python decorator="dataclass(frozen=True)"> = {
```
Comment (aryx) [inline interfaces/Output_from_core.atd:246]: can we make this the default?

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -21,6 +21,7 @@ repos:
     rev: 22.3.0
     hooks:
       - id: black
+        args: [--exclude=/semgrep/semgrep/output_from_core.py]
```
Comment (aryx) [inline .pre-commit-config.yaml:24]: I don't think this works. I've regenerated a few times this file and it got reformatted by some tools (black I guess).

File: interfaces/Output_from_core.atd (hunk 3 above)
Comment (mjambon) [reply interfaces/Output_from_core.atd:246]: please open an issue: https://github.com/ahrefs/atd/issues

Comment (aryx) [reply interfaces/Output_from_core.atd:246]: done: https://github.com/ahrefs/atd/issues/289

**Pull request merged**